src/seraphis_mocks/scan_context_async_mock.h

```diff
@@ -0,0 +1,174 @@
+{
+namespace mocks
+{
+//-------------------------------------------------------------------------------------------------------------------
+struct ChunkRequest final
+{
+    std::uint64_t start_index;
+    std::uint64_t requested_chunk_size;
+    bool gap_fill;
+};
+
+struct PendingChunk final
+{
+    ChunkRequest chunk_request;
+    sp::scanning::PendingChunkContext pending_context;
+    sp::scanning::PendingChunkData pending_data;
+};
+
+static inline bool operator<(const PendingChunk &pending_chunk1, const PendingChunk &pending_chunk2)
+{
+    return pending_chunk1.chunk_request.start_index < pending_chunk2.chunk_request.start_index;
+}
+
+typedef struct {
+    cryptonote::transaction tx;
+    crypto::hash tx_hash;
+    uint64_t total_output_count_before_tx;
+} parsed_transaction_t;
+
+typedef struct {
+    uint64_t block_index;
+    uint64_t timestamp;
+    crypto::hash block_hash;
+    crypto::hash prev_block_hash;
+    std::vector<parsed_transaction_t> parsed_txs;
+} parsed_block_t;
+
+typedef struct {
+    rct::key tx_hash;
+    uint64_t block_index;
+    uint64_t timestamp;
+    uint64_t total_output_count_before_tx;
+    uint64_t unlock_time;
+    sp::TxExtra tx_extra;
+    std::vector<sp::LegacyEnoteVariant> enotes;
+    std::vector<crypto::key_image> legacy_key_images;
+} legacy_transaction_to_scan_t;
+//-------------------------------------------------------------------------------------------------------------------
+////
+// WARNING: if the chunk size increment exceeds the max chunk size obtainable from the raw chunk data source, then
+//          this will be less efficient because it will need to 'gap fill' continuously
+///
+class AsyncScanContextLegacy final : public ScanContextLedger
+{
+public:
+    AsyncScanContextLegacy(const std::uint64_t pending_chunk_max_queue_size,
+            const std::uint64_t chunk_size_increment,
+            const std::uint64_t max_get_blocks_attempts,
+            sp::mocks::EnoteFindingContextMockLegacy &enote_finding_context) :
+        m_pending_chunk_max_queue_size{pending_chunk_max_queue_size},
+        m_max_scanning_queue_size{pending_chunk_max_queue_size},
+        m_chunk_size_increment{chunk_size_increment},
+        m_max_get_blocks_attempts{max_get_blocks_attempts},
+        m_enote_finding_context{enote_finding_context},
+        m_pause_pending_queue{false}
+    {
+        assert(m_pending_chunk_max_queue_size > 0);
+        assert(m_chunk_size_increment > 0);
+        assert(m_max_get_blocks_attempts > 0);
+    }
+
+    ~AsyncScanContextLegacy()
+    {
+        std::unique_lock<std::mutex> lock{m_pending_queue_mutex};
+        wait_until_pending_queue_clears(lock);
+    }
+
+    void begin_scanning_from_index(const std::uint64_t start_index, const std::uint64_t max_chunk_size);
+
+    std::unique_ptr<sp::scanning::LedgerChunk> get_onchain_chunk();
+
+    /// stop the current scanning process (should be no-throw no-fail)
+    void terminate_scanning() override { /* no-op */ }
```

Comment:
Are there plans to make it stoppable in the future?

Reply:
If the wallet is scanning and the user changes their restore height or wants to use a different wallet, that should cleanly terminate the scanner